### PR TITLE
Statically assert sizeof(J9Class) == sizeof(J9ArrayClass)

### PR DESCRIPTION
**Pull request by DanHeidinga**
After seeing https://github.com/eclipse/openj9/pull/7677, I tried adding a static assert for it and found a second problem.

This commit fixes the problem and adds the assert.

Signed-off-by: Dan Heidinga <daniel_heidinga@ca.ibm.com>

### CODE REVIEW

File: runtime/oti/j9nonbuilder.h [hunk 1]
```diff
@@ -3082,15 +3082,21 @@ typedef struct J9ArrayClass {
 	struct J9Method** specialSplitMethodTable;
 	struct J9JITExceptionTable* jitMetaDataList;
 	struct J9Class* gcLink;
+	struct J9Class* hostClass;
 #if defined(J9VM_OPT_VALHALLA_NESTMATES)
 	struct J9Class* nestHost;
 #endif /* defined(J9VM_OPT_VALHALLA_NESTMATES) */
-#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 	/* Added temporarily for consistency */
```
Comment (XxAdi101xX) [inline runtime/oti/j9nonbuilder.h]: could you remove this comment as well, just to prevent confusion

**Comment by dsouzai** [comment]
I took a look at the differences between J9Class and J9ArrayClass and it's just these fields
```
	struct J9Method* ramMethods;
	UDATA* ramStatics;
	...
	UDATA totalInstanceSize;
```
```
	struct J9Class* leafComponentType;
	UDATA arity;
	...
	struct J9Class* componentType;
```

Not suggesting it needs to be done in this PR, but really those three fields should just be unions in `J9Class` and then we do a `typedef J9Class J9ArrayClass`. The size being the same doesn't prevent someone from accidentally re-ordering the fields. all uses of, for example `->ramMethods` would be replaced with `->methodOrClass.ramMethods`.

I'd be willing to do the search and replace at some point if you're ok with that idea; I've been meaning to get rid of `J9AOTConfig` at some point anyway, might as well do it all at one time.

**Comment by gacholio** [comment]
Have you tried this with xlC? From what I've seen, it simply doesn't support static_assert at all.

**Comment by DanHeidinga** [comment]
Jenkins compile plinux,aix jdk8

**Comment by gacholio** [comment]
The union is a good idea - the separate structs are a legacy from builder when this stuff was easily autogenerated.

**Comment by gacholio** [comment]
```
09:24:40  "../oti/j9nonbuilder.h", line 3096.1: 1540-2935 (S) A C++0x feature is being used. Use the extended0x langlvl or langlvl "Static_assert" to enable this feature.
09:24:40  "../oti/j9nonbuilder.h", line 3096.1: 1540-2955 (I) Compilation may require C++0x Feature. Try using the option "-qlanglvl=static_assert".
```

**Comment by DanHeidinga** [comment]
Jenkins compile plinux,aix,osx,win,xlinux jdk8

**Comment by DanHeidinga** [comment]
Pushed a change to limit the static assert to x86 only.  As this is the primary dev env for many of the committers, the static assert will catch any breakage here even if it's not enabled everywhere.

**Comment by DanHeidinga** [comment]
Going to try limiting this to linux x86 to still get some benefit from the static assert while our compilers don't all support it.

**Comment by DanHeidinga** [comment]
Jenkins compile plinux,osx,win,xlinux jdk8

**Comment by DanHeidinga** [comment]
@gacholio Not the prettiest fix but it does get the static_assert enabled on linux x86 so we should generally catch any mismatches.

**Comment by gacholio** [comment]
jenkins compile aix

**Comment by DanHeidinga** [comment]
jenkins compile aix jdk8

**Comment by fjeremic** [comment]
> Have you tried this with xlC? From what I've seen, it simply doesn't support static_assert at all.

It does work. There are `static_asserts` being used in the JIT for quite some time:
https://github.com/eclipse/openj9/blob/5a3774c9e698f1e0661420efebc0d93dd6075d1e/runtime/compiler/il/J9IL.cpp#L194-L225

The reason there may be troubles in the VM is because the VM and JIT build with different build flags. The VM can use `static_assert` as well. All we need to do is add the option which is used in the JIT already:
https://github.com/eclipse/openj9/blob/5a3774c9e698f1e0661420efebc0d93dd6075d1e/runtime/compiler/build/toolcfg/zos-xlc/common.mk#L82-L83

I've opened up an issue to track this work:

**Pull request merged**